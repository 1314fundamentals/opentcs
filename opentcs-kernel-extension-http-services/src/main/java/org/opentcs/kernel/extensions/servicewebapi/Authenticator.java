// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import static java.util.Objects.requireNonNull;

import com.google.common.base.Strings;
import io.javalin.http.Context;
import jakarta.inject.Inject;
import java.util.Objects;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Authenticates incoming requests.
 */
public class Authenticator {

  /**
   * This class's logger.
   */
  private static final Logger LOG = LoggerFactory.getLogger(Authenticator.class);
  /**
   * Defines the required access rules.
   */
  private final ServiceWebApiConfiguration configuration;

  /**
   * Creates a new instance.
   *
   * @param configuration Defines the required access rules.
   */
  @Inject
  public Authenticator(ServiceWebApiConfiguration configuration) {
    this.configuration = requireNonNull(configuration, "configuration");
  }

  /**
   * Checks whether authentication is required and the given request is authenticated.
   *
   * @param context The context to be checked.
   * @return <code>true</code> if, and only if, authentication is required and the given request is
   * authenticated.
   */
  public boolean isAuthenticated(Context context) {
    requireNonNull(context, "context");

    String requestAccessKey = context.header(HttpConstants.HEADER_NAME_ACCESS_KEY);
    LOG.debug(
        "Provided access key in header is '{}', required value is '{}'",
        requestAccessKey,
        configuration.accessKey()
    );

    // Any empty access key indicates authentication is not required.
    if (Strings.isNullOrEmpty(configuration.accessKey())) {
      LOG.debug("No access key, authentication not required.");
      return true;
    }

    return Objects.equals(requestAccessKey, configuration.accessKey());
  }

}
