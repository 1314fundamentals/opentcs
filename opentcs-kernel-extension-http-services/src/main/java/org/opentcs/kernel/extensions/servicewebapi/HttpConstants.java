// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

/**
 * Defines some HTTP-related constants.
 */
public class HttpConstants {

  /**
   * Name of the header that is expected to contain the API access keys.
   */
  public static final String HEADER_NAME_ACCESS_KEY = "X-Api-Access-Key";
  /**
   * Content type for plain text.
   */
  public static final String CONTENT_TYPE_TEXT_PLAIN_UTF8 = "text/plain; charset=utf-8";
  /**
   * Content type for JSON structures.
   */
  public static final String CONTENT_TYPE_APPLICATION_JSON_UTF8 = "application/json; charset=utf-8";

  /**
   * Prevents instantiation.
   */
  private HttpConstants() {
  }
}
