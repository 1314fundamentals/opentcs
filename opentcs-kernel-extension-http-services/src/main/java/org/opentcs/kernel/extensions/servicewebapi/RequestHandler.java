// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi;

import io.javalin.apibuilder.EndpointGroup;
import org.opentcs.components.Lifecycle;

/**
 * A request handler.
 */
public interface RequestHandler
    extends
      Lifecycle {

  /**
   * Creates the handler's routes for the web service.
   */
  EndpointGroup createRoutes();
}
