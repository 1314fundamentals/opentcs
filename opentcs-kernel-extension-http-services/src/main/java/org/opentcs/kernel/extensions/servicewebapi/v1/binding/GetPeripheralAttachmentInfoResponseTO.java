// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import static java.util.Objects.requireNonNull;

import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;
import java.util.List;
import java.util.stream.Collectors;
import org.opentcs.drivers.peripherals.management.PeripheralAttachmentInformation;

/**
 */
public class GetPeripheralAttachmentInfoResponseTO {

  @Nonnull
  private String locationName;

  @Nonnull
  private List<String> availableCommAdapters;

  @Nonnull
  private String attachedCommAdapter;

  public GetPeripheralAttachmentInfoResponseTO(
      @Nonnull
      String locationName,
      @Nonnull
      String attachedCommAdapter,
      @Nonnull
      List<String> availableCommAdapters
  ) {
    this.locationName = requireNonNull(locationName, "locationName");
    this.attachedCommAdapter = requireNonNull(attachedCommAdapter, "attachedCommAdapter");
    this.availableCommAdapters = requireNonNull(availableCommAdapters, "availableCommAdapters");
  }

  @Nonnull
  public String getLocationName() {
    return locationName;
  }

  public GetPeripheralAttachmentInfoResponseTO setLocationName(
      @Nonnull
      String locationName
  ) {
    this.locationName = requireNonNull(locationName, "locationName");
    return this;
  }

  @Nonnull
  public List<String> getAvailableCommAdapters() {
    return availableCommAdapters;
  }

  public GetPeripheralAttachmentInfoResponseTO setAvailableCommAdapters(
      @Nonnull
      List<String> availableCommAdapters
  ) {
    this.availableCommAdapters = requireNonNull(availableCommAdapters, "availableCommAdapters");
    return this;
  }

  @Nonnull
  public String getAttachedCommAdapter() {
    return attachedCommAdapter;
  }

  public GetPeripheralAttachmentInfoResponseTO setAttachedCommAdapter(
      @Nonnull
      String attachedCommAdapter
  ) {
    this.attachedCommAdapter = requireNonNull(attachedCommAdapter, "attachedCommAdapter");
    return this;
  }

  public static GetPeripheralAttachmentInfoResponseTO fromAttachmentInformation(
      @Nullable
      PeripheralAttachmentInformation peripheralAttachmentInfo
  ) {
    if (peripheralAttachmentInfo == null) {
      return null;
    }

    List<String> availableAdapters = peripheralAttachmentInfo.getAvailableCommAdapters()
        .stream()
        .map(description -> description.getClass().getName())
        .collect(Collectors.toList());

    return new GetPeripheralAttachmentInfoResponseTO(
        peripheralAttachmentInfo.getLocationReference().getName(),
        peripheralAttachmentInfo.getAttachedCommAdapter().getClass().getName(),
        availableAdapters
    );
  }
}
