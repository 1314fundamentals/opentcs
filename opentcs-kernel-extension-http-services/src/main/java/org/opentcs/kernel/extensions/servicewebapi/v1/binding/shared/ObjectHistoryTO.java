// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared;

import static java.util.Objects.requireNonNull;

import java.time.Instant;
import java.util.List;

/**
 */
public class ObjectHistoryTO {
  private List<ObjectHistoryEntryTO> entries;

  public ObjectHistoryTO(
      List<ObjectHistoryEntryTO> entries
  ) {
    this.entries = requireNonNull(entries, "entries");
  }

  public List<ObjectHistoryEntryTO> getEntries() {
    return entries;
  }

  public ObjectHistoryTO setEntries(List<ObjectHistoryEntryTO> entries) {
    this.entries = entries;
    return this;
  }

  /**
   */
  public static class ObjectHistoryEntryTO {
    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;

    public ObjectHistoryEntryTO(
        Instant timeStamp,
        String eventCode,
        List<String> supplements
    ) {
      this.timestamp = requireNonNull(timeStamp, "timeStamp");
      this.eventCode = requireNonNull(eventCode, "eventCode");
      this.supplements = requireNonNull(supplements, "supplements");
    }

    public Instant getTimestamp() {
      return timestamp;
    }

    public ObjectHistoryEntryTO setTimestamp(Instant timestamp) {
      this.timestamp = timestamp;
      return this;
    }

    public String getEventCode() {
      return eventCode;
    }

    public ObjectHistoryEntryTO setEventCode(String eventCode) {
      this.eventCode = eventCode;
      return this;
    }

    public List<String> getSupplements() {
      return supplements;
    }

    public ObjectHistoryEntryTO setSupplements(List<String> supplements) {
      this.supplements = supplements;
      return this;
    }
  }
}
