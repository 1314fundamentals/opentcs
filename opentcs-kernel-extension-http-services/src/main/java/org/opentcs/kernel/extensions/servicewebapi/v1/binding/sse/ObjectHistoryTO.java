// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.ObjectHistory;

/**
 * A transfer object (to be used with the SSE API) representing an {@link ObjectHistory} instance.
 */
// CHECKSTYLE:OFF
@NoArgsConstructor
@Getter
@Setter
@Accessors(chain = true)
public class ObjectHistoryTO {

  private List<ObjectHistoryEntryTO> entries;

  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class ObjectHistoryEntryTO {

    private Instant timestamp;
    private String eventCode;
    private List<String> supplements;
  }
}
// CHECKSTYLE:ON
