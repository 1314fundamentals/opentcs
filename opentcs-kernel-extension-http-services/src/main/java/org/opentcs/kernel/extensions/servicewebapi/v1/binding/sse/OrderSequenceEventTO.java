// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import javax.annotation.Nullable;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.experimental.Accessors;
import org.opentcs.data.TCSObjectEvent;

/**
 * A transfer object (to be used with the SSE API) representing a {@link TCSObjectEvent} for an
 * order sequence.
 */
public class OrderSequenceEventTO {

  private final OrderSequenceTO currentObjectState;
  private final OrderSequenceTO previousObjectState;

  public OrderSequenceEventTO(
      @Nullable
      OrderSequenceTO currentObjectState,
      @Nullable
      OrderSequenceTO previousObjectState
  ) {
    this.currentObjectState = currentObjectState;
    this.previousObjectState = previousObjectState;
  }

  public OrderSequenceTO getCurrentObjectState() {
    return currentObjectState;
  }

  public OrderSequenceTO getPreviousObjectState() {
    return previousObjectState;
  }

  // CHECKSTYLE:OFF
  @NoArgsConstructor
  @Getter
  @Setter
  @Accessors(chain = true)
  public static class OrderSequenceTO {

    private String name;
    private Map<String, String> properties;
    private ObjectHistoryTO history;
    private String type;
    private List<String> orders;
    private int finishedIndex;
    private boolean complete;
    private boolean finished;
    private boolean failureFatal;
    private String intendedVehicle;
    private String processingVehicle;
    private Instant creationTime;
    private Instant finishedTime;
  }
  // CHECKSTYLE:ON
}
