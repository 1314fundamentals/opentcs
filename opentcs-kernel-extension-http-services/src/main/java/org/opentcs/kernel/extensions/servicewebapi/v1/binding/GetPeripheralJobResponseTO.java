// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.binding;

import java.time.Instant;
import java.util.List;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.plantmodel.PeripheralJobStateTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.PeripheralOperationDescription;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.shared.Property;

/**
 * The current state of a peripheral job.
 */
public class GetPeripheralJobResponseTO {

  private String name;

  private String reservationToken;

  private String relatedVehicle;

  private String relatedTransportOrder;

  private PeripheralOperationDescription peripheralOperation;

  private PeripheralJobStateTO state;

  private Instant creationTime;

  private Instant finishedTime;

  private List<Property> properties;

  public GetPeripheralJobResponseTO() {
  }

  public String getName() {
    return name;
  }

  public GetPeripheralJobResponseTO setName(String name) {
    this.name = name;
    return this;
  }

  public String getReservationToken() {
    return reservationToken;
  }

  public GetPeripheralJobResponseTO setReservationToken(String reservationToken) {
    this.reservationToken = reservationToken;
    return this;
  }

  public String getRelatedVehicle() {
    return relatedVehicle;
  }

  public GetPeripheralJobResponseTO setRelatedVehicle(String relatedVehicle) {
    this.relatedVehicle = relatedVehicle;
    return this;
  }

  public String getRelatedTransportOrder() {
    return relatedTransportOrder;
  }

  public GetPeripheralJobResponseTO setRelatedTransportOrder(String relatedTransportOrder) {
    this.relatedTransportOrder = relatedTransportOrder;
    return this;
  }

  public PeripheralOperationDescription getPeripheralOperation() {
    return peripheralOperation;
  }

  public GetPeripheralJobResponseTO setPeripheralOperation(
      PeripheralOperationDescription peripheralOperation
  ) {
    this.peripheralOperation = peripheralOperation;
    return this;
  }

  public PeripheralJobStateTO getState() {
    return state;
  }

  public GetPeripheralJobResponseTO setState(PeripheralJobStateTO state) {
    this.state = state;
    return this;
  }

  public Instant getCreationTime() {
    return creationTime;
  }

  public GetPeripheralJobResponseTO setCreationTime(Instant creationTime) {
    this.creationTime = creationTime;
    return this;
  }

  public Instant getFinishedTime() {
    return finishedTime;
  }

  public GetPeripheralJobResponseTO setFinishedTime(Instant finishedTime) {
    this.finishedTime = finishedTime;
    return this;
  }

  public List<Property> getProperties() {
    return properties;
  }

  public GetPeripheralJobResponseTO setProperties(List<Property> properties) {
    this.properties = properties;
    return this;
  }
}
