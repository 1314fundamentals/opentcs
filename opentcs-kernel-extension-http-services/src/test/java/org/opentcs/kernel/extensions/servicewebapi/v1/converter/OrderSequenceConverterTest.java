// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

import java.time.Instant;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.OrderSequence;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.GetOrderSequenceResponseTO;

/**
 * Tests for {@link OrderSequenceConverter}.
 */
public class OrderSequenceConverterTest {
  private final OrderSequenceConverter converter = new OrderSequenceConverter();

  @Test
  public void checkToGetOrderSequenceResponseTO() {
    OrderSequence sequence = new OrderSequence("OS1")
        .withOrder(new TransportOrder("T1", List.of()).getReference())
        .withComplete(true)
        .withFailureFatal(true)
        .withFinished(false)
        .withFinishedIndex(0)
        .withCreationTime(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH)
        .withType("some-type")
        .withIntendedVehicle(new Vehicle("V1").getReference())
        .withProcessingVehicle(new Vehicle("V2").getReference())
        .withProperties(Map.of("some-key", "some-value"));

    GetOrderSequenceResponseTO response = converter.toGetOrderSequenceResponseTO(sequence);

    assertThat(response.getName(), is("OS1"));
    assertThat(response.isComplete(), is(true));
    assertThat(response.isFailureFatal(), is(true));
    assertThat(response.isFinished(), is(false));
    assertThat(response.getFinishedIndex(), is(0));
    assertThat(response.getCreationTime(), is(Instant.EPOCH));
    assertThat(response.getFinishedTime(), is(Instant.EPOCH));
    assertThat(response.getType(), is("some-type"));
    assertThat(response.getIntendedVehicle(), is("V1"));
    assertThat(response.getOrders().size(), is(1));
    assertThat(response.getOrders().getFirst(), is("T1"));
    assertThat(response.getProperties().size(), is(1));
    assertThat(response.getProperties().getFirst().getKey(), is("some-key"));
    assertThat(response.getProperties().getFirst().getValue(), is("some-value"));
  }
}
