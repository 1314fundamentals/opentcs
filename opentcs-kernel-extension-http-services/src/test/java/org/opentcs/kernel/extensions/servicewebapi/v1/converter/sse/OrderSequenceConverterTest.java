// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.kernel.extensions.servicewebapi.v1.converter.sse;

import static org.assertj.core.api.Assertions.assertThat;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.assertj.core.groups.Tuple;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.data.ObjectHistory;
import org.opentcs.data.model.Vehicle;
import org.opentcs.data.order.OrderSequence;
import org.opentcs.data.order.TransportOrder;
import org.opentcs.kernel.extensions.servicewebapi.JsonBinder;
import org.opentcs.kernel.extensions.servicewebapi.TimestampScrubber;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.ObjectHistoryTO;
import org.opentcs.kernel.extensions.servicewebapi.v1.binding.sse.OrderSequenceEventTO;

/**
 * Tests for {@link OrderSequenceConverter}.
 */
class OrderSequenceConverterTest {

  private JsonBinder jsonBinder;
  private OrderSequenceConverter converter;

  @BeforeEach
  void setUp() {
    jsonBinder = new JsonBinder();
    converter = new OrderSequenceConverter();
  }

  @Test
  void convert() {
    OrderSequence orderSequence = new OrderSequence("sequence-1")
        .withProperties(new TreeMap<>(Map.of("key-1", "value-1", "key-2", "value-2")))
        .withHistoryEntry(
            new ObjectHistory.Entry(Instant.EPOCH, "event-code-1", List.of("supplement-1"))
        )
        .withType("type-1")
        .withOrder(new TransportOrder("order-1", List.of()).getReference())
        .withOrder(new TransportOrder("order-2", List.of()).getReference())
        .withFinishedIndex(1)
        .withComplete(true)
        .withFinished(true)
        .withFailureFatal(true)
        .withIntendedVehicle(new Vehicle("vehicle-1").getReference())
        .withProcessingVehicle(new Vehicle("vehicle-2").getReference())
        .withCreationTime(Instant.EPOCH)
        .withFinishedTime(Instant.EPOCH);

    OrderSequenceEventTO.OrderSequenceTO result = converter.convert(orderSequence);

    Approvals.verify(
        jsonBinder.toJson(result),
        // Scrub timestamps in the result's JSON representation as some of them may be generated
        // during test runtime, making it impossible to use ApprovalTests.
        new Options(new TimestampScrubber())
    );
  }

  /**
   * Ensures that timestamps are converted correctly. (Testing this is necessary, since timestamps
   * are scrubbed in the ApprovalTest and therefore their conversion is not covered there.)
   */
  @Test
  void convertsTimestampsCorrectly() {
    Instant time = Instant.parse("2025-07-22T12:03:00Z");
    OrderSequence orderSequence = new OrderSequence("sequence-1")
        .withHistoryEntry(
            new ObjectHistory.Entry(
                time.plus(2, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        )
        .withCreationTime(time)
        .withFinishedTime(time.plus(1, ChronoUnit.HOURS));

    OrderSequenceEventTO.OrderSequenceTO result = converter.convert(orderSequence);

    assertThat(result.getCreationTime()).isEqualTo(time);
    assertThat(result.getFinishedTime()).isEqualTo(time.plus(1, ChronoUnit.HOURS));
    assertThat(result.getHistory().getEntries())
        .extracting(
            ObjectHistoryTO.ObjectHistoryEntryTO::getTimestamp,
            ObjectHistoryTO.ObjectHistoryEntryTO::getEventCode,
            ObjectHistoryTO.ObjectHistoryEntryTO::getSupplements
        )
        .contains(
            Tuple.tuple(
                time.plus(2, ChronoUnit.HOURS),
                "event-code-1",
                List.of("supplement-1")
            )
        );
  }
}
