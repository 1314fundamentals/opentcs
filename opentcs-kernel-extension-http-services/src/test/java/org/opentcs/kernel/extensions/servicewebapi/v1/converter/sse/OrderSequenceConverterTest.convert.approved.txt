{
  "name" : "sequence-1",
  "properties" : {
    "key-1" : "value-1",
    "key-2" : "value-2"
  },
  "history" : {
    "entries" : [ {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceCreated",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "event-code-1",
      "supplements" : [ "supplement-1" ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceOrderAppended",
      "supplements" : [ "order-1" ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceOrderAppended",
      "supplements" : [ "order-2" ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceCompleted",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceFinished",
      "supplements" : [ ]
    }, {
      "timestamp" : "[Timestamp]",
      "eventCode" : "tcs:history:sequenceProcVehicleChanged",
      "supplements" : [ "vehicle-2" ]
    } ]
  },
  "type" : "type-1",
  "orders" : [ "order-1", "order-2" ],
  "finishedIndex" : 1,
  "complete" : true,
  "finished" : true,
  "failureFatal" : true,
  "intendedVehicle" : "vehicle-1",
  "processingVehicle" : "vehicle-2",
  "creationTime" : "[Timestamp]",
  "finishedTime" : "[Timestamp]"
}